### PR TITLE
test: Add format-dependent comparison to bctest

### PR DESCRIPTION
**Pull request by laanwj**
This splits the output comparison for `bitcoin-tx` into two steps:
- First, check for data mismatch, parsing the data as json or hex depending on the extension of the output file
- Then, check if the literal string matches

For either of these cases give a different error.

This prevents wild goose chases when e.g. a trailing space doesn't match exactly (by helping narrow down the problem), and makes sure that both test output and examples are valid data of the purported format.

### CODE REVIEW

**Comment by JeremyRubin** [comment]
utAck, seems useful!

**Comment by jnewbery** [comment]
Yes, makes sense. I should have included something like this when I added the json tests.

I think https://github.com/bitcoin/bitcoin/pull/9023 (which outputs a contextual diff for failing test cases) should make it much easier to track down formatting errors, since it'll point you to the exact line that's different.

However, I think this is still useful for verifying that the output is indeed valid json.

Only nit is that `compare_output()` can raise errors which aren't caught anywhere and don't result in useful logging. For example, here's the logging if the output file is invalid json:

```
~/bitcoin/src$ test/bitcoin-util-test.py --srcdir='.'
Traceback (most recent call last):
  File "test/bitcoin-util-test.py", line 32, in <module>
    bctest.bctester(srcdir + "/test/data", "bitcoin-util-test.json", buildenv, verbose = verbose)
  File "/home/vagrant/bitcoin/src/test/bctest.py", line 72, in bctester
    bctest(testDir, testObj, buildenv.exeext)
  File "/home/vagrant/bitcoin/src/test/bctest.py", line 50, in bctest
    if not compare_output(outs[0], outputData, outputType):
  File "/home/vagrant/bitcoin/src/test/bctest.py", line 14, in compare_output
    match = (json.loads(a) == json.loads(b))
  File "/usr/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
    obj, end = self.scan_once(s, idx)
ValueError: Invalid control character at: line 3 column 14 (char 95)
```

It'd be nice to wrap the json calls in try/except, and maybe have the final else branch of `compare_output()` print an error and exit rather than raise an error.

**Comment by laanwj** [comment]
Yea, that error 'Invalid control character at: line 3 column 14 (char 95)" is actually very good and specific. The only thing missing is context of what file it occurs in. Would make sense to print it with the filename.

File: src/test/bctest.py [hunk 1]
```diff
@@ -6,6 +6,17 @@
...
 import json
 import sys
+import binascii
+
+def compare_output(a, b, fmt):
+	'''Data-format dependent output comparison.'''
+	if fmt == 'json': # json: compare parsed data
+		match = (json.loads(a) == json.loads(b))
+	elif fmt == 'hex': # hex: parse and compare binary data
+		match = (binascii.a2b_hex(a.rstrip()) == binascii.a2b_hex(b.rstrip()))
```
Comment (JeremyRubin) [inline src/test/bctest.py]: Why is this rstripped rather than stripped? Is there a case where you want to fail here if leading whitespace differs?

Comment (laanwj) [inline src/test/bctest.py]: Leading whitespace is much harder to accidentally introduce than trailing whitespace (e.g. trailing spaces, newlines etc). But I'm fine with changing this to strip() instead.

**Comment by laanwj** [comment]
- Improved error reporting:
```
Error parsing expected output blanktx.json as json: Expecting property name: line 4 column 18 (char 179)
Error parsing expected output blanktx.hex as hex: Non-hexadecimal digit found
Error parsing command output as hex: Non-hexadecimal digit found
```
- Replaced rstrip with strip

**Comment by jnewbery** [comment]
Looks good. Tested ACK

**Comment by MarcoFalke** [comment]
utACK 6c5cd9d

**Pull request merged**